lib/vec.c

```diff
@@ -178,13 +186,19 @@ static int bufvec_alloc(struct m0_bufvec *bufvec,
 
 			if (bufvec->ov_buf[i] == NULL)
 				goto fail;
+			if (M0_FI_ENABLED("buf-alloc-fail"))
+				goto fail;
 			bufvec->ov_vec.v_count[i] = seg_size;
 		}
 	}
 	return 0;
 
 fail:
-	m0_bufvec_free(bufvec);
+	if (seg_size != 0)
+		m0_bufvec_free(bufvec);
+	else
+		m0_bufvec_free2(bufvec);
+
```

Comment:
The original code still works, right?

Because m0_bufvec_free() tries to free the buffers. But the buffers may not have been allocated successfully. So, the m0_free() operation is a nil operation.